### DISCUSSION TITLE
performance, file size issue

### DISCUSSION THREAD

**Discussion by pavlexander** [G-Research/ParquetSharp#377] [Q&A]
Hi!

Just started using this library and trying to compare it's performance to main competitor Parquet.NET (which is also referenced on a main page)!

## the data

```csharp
    public class SingleCandlestickMP
    {
        public long Id { get; set; }
        public long OpenTime { get; set; }
        public long Open { get; set; }
        public long High { get; set; }
        public long Low { get; set; }
        public long Close { get; set; }
        public long Volume { get; set; }
        public long CloseTime { get; set; }
    }
```

**2_673_685 number of rows**

## parquetSharp data save approaches

I have so far managed to save the data using 2 approaches that seemingly yield the same results.

### approach 1:

```csharp
        public static async Task SaveToParquet2(this IEnumerable<SingleCandlestickMP> data, string fileFullPath)
        {
            var columns = new Column[]
            {
                new Column<long>("Id"),
                new Column<long>("OpenTime"),
                new Column<long>("Open"),
                new Column<long>("High"),
                new Column<long>("Low"),
                new Column<long>("Close"),
                new Column<long>("Volume"),
                new Column<long>("CloseTime")
            };

            var rows = data.ToArray();

            using var writer = ParquetFile.CreateRowWriter<SingleCandlestickMP>(fileFullPath, columns, Compression.Snappy);
            writer.WriteRows(rows);
            writer.Close();
        }
```

### approach 2

```csharp
        public static async Task SaveToParquet2(this IEnumerable<SingleCandlestickMP> data, string fileFullPath)
        {
            var columns = new Column[]
            {
                new Column<long>("Id"),
                new Column<long>("OpenTime"),
                new Column<long>("Open"),
                new Column<long>("High"),
                new Column<long>("Low"),
                new Column<long>("Close"),
                new Column<long>("Volume"),
                new Column<long>("CloseTime")
            };

            var i = data.Select(x => x.Id).ToArray();
            var ot = data.Select(x => x.OpenTime).ToArray();
            var o = data.Select(x => x.Open).ToArray();
            var h = data.Select(x => x.High).ToArray();
            var l = data.Select(x => x.Low).ToArray();
            var c = data.Select(x => x.Close).ToArray();
            var v = data.Select(x => x.Volume).ToArray();
            var ct = data.Select(x => x.CloseTime).ToArray();*/

        using (var stream = new FileStream(fileFullPath, FileMode.Create))
        {
            using var writer = new ManagedOutputStream(stream);
            using var fileWriter = new ParquetFileWriter(writer, columns);

            using RowGroupWriter rowGroup = fileWriter.AppendRowGroup();

            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(i);
            }
            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(ot);
            }
            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(o);
            }
            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(h);
            }
            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(l);
            }
            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(c);
            }
            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(v);
            }
            using (var objectIdWriter = rowGroup.NextColumn().LogicalWriter<long>())
            {
                objectIdWriter.WriteBatch(ct);
            }
            }
        }
```

### quick summary and a side question

personally I don't see much use in approach 2, since it's more complex and more error prone. Is there something I gain from it when compared to approach 1?

##  parquet.net approach

Regardless, moving on back to the main topic. Here is the code I am using to save the data in parquet.net

```csharp
        public static async Task SaveToParquet(this IEnumerable<SingleCandlestickMP> data, string fileFullPath)
        {
            await ParquetSerializer.SerializeAsync(data, fileFullPath, new ParquetSerializerOptions()
            {
                CompressionMethod = Parquet.CompressionMethod.Snappy
            });
        }
```

see that in all 3 approaches Snappy compression method is used

##  results

* time takes to save the file `ParquetSharp`: ~800ms (817). **File size: 81gb (approach 1)**
* time takes to save the file `Parquet.net`: ~800ms (788). **File size: 31gb.**

##  results interpretation

As you see both libraries take almost the same amount of time to save the file. This is OK.
**But then the main issue is the file size.** Is there a reason why `ParquetSharp` result takes that much space?

I have extracted the metadata with `ParquetViewer` and compared both files. It seems like there are quite a lot of differences. Some of them are (left: `parquet.net`, right: `parquetSharp`):

schema version:

![image](https://github.com/G-Research/ParquetSharp/assets/20644772/df9e6c39-87f2-45a3-85a1-f33857a0818e)

encodings:

![image](https://github.com/G-Research/ParquetSharp/assets/20644772/19feb822-1ee2-494b-b03d-91b8ba86dbf6)

something else

![image](https://github.com/G-Research/ParquetSharp/assets/20644772/3bc253a8-0626-4b70-b84e-169a67b0a82d)

### package versions

* `<PackageReference Include="Parquet.Net" Version="4.16.4" />` (Parquet.Net version 4.16.4 (build 42339e08d7520ef1301b27689e1d1c02d91b058e))
* `<PackageReference Include="ParquetSharp" Version="12.0.1" />` (parquet-cpp-arrow version 12.0.1)

**Comment by marcin-krystianc** [comment]
Hi @pavlexander,

I think the factor to consider is default column encodings used by `Parquet.Net` and `ParquetSharp`. 
Try disabling dictionary encoding (`ParquetSharp`) and enabling `DeltaBinaryPacked` encoding and see whether it improved the storage size: 
```
     var builder = new ParquetSharp.WriterPropertiesBuilder();
     builder.DisableDictionary();
     builder.Encoding(ParquetSharp.Encoding.DeltaBinaryPacked);
```

**Reply by pavlexander** [reply]
Hey @marcin-krystianc 

First of all I would like to apologize about the file sizes. They are not in `Gbs` but in `Mbs` :)

That aside -  I have tried your suggested solution and see some improvements, but still the file size was not as low as I would prefer.. So I decided to do an extensive test of all possible compression levels, encodings and disabled dictionary flag.

In the results below is the aggregated list of combinations that produced any kind of output file. There were a bunch of combinations that threw exceptions as well that I have removed from this list..

The combination consists of 3 parameters: `Compression`,`Encoding`,`isDictionaryDisabled` followed by elapsed time and resulting file size

```txt
Uncompressed,Plain,True Elapsed: 468 ms, Size: 163.2 mb
Uncompressed,Plain,False Elapsed: 426 ms, Size: 106.63 mb
Snappy,Plain,True Elapsed: 531 ms, Size: 83.9 mb
Snappy,Plain,False Elapsed: 614 ms, Size: 81.15 mb
Gzip,Plain,True Elapsed: 7167 ms, Size: 49.05 mb
Gzip,Plain,False Elapsed: 6380 ms, Size: 50.75 mb
Brotli,Plain,True Elapsed: 10159 ms, Size: 44.54 mb
Brotli,Plain,False Elapsed: 9312 ms, Size: 46.59 mb
Zstd,Plain,True Elapsed: 572 ms, Size: 57.63 mb
Zstd,Plain,False Elapsed: 600 ms, Size: 59.14 mb
Lz4,Plain,True Elapsed: 406 ms, Size: 89.11 mb
Lz4,Plain,False Elapsed: 633 ms, Size: 81.27 mb
Lz4Hadoop,Plain,True Elapsed: 552 ms, Size: 89.11 mb
Lz4Hadoop,Plain,False Elapsed: 517 ms, Size: 81.27 mb
Uncompressed,Rle,False Elapsed: 418 ms, Size: 106.63 mb
Snappy,Rle,False Elapsed: 749 ms, Size: 81.15 mb
Gzip,Rle,False Elapsed: 6356 ms, Size: 50.75 mb
Brotli,Rle,False Elapsed: 9448 ms, Size: 46.59 mb
Zstd,Rle,False Elapsed: 608 ms, Size: 59.14 mb
Lz4,Rle,False Elapsed: 536 ms, Size: 81.27 mb
Lz4Hadoop,Rle,False Elapsed: 532 ms, Size: 81.27 mb
Uncompressed,BitPacked,False Elapsed: 534 ms, Size: 106.63 mb
Snappy,BitPacked,False Elapsed: 583 ms, Size: 81.15 mb
Gzip,BitPacked,False Elapsed: 6501 ms, Size: 50.75 mb
Brotli,BitPacked,False Elapsed: 9371 ms, Size: 46.59 mb
Zstd,BitPacked,False Elapsed: 589 ms, Size: 59.14 mb
Lz4,BitPacked,False Elapsed: 634 ms, Size: 81.27 mb
Lz4Hadoop,BitPacked,False Elapsed: 623 ms, Size: 81.27 mb
Uncompressed,DeltaBinaryPacked,True Elapsed: 300 ms, Size: 102.48 mb
Uncompressed,DeltaBinaryPacked,False Elapsed: 371 ms, Size: 106.63 mb
Snappy,DeltaBinaryPacked,True Elapsed: 597 ms, Size: 39.25 mb
Snappy,DeltaBinaryPacked,False Elapsed: 555 ms, Size: 81.15 mb
Gzip,DeltaBinaryPacked,True Elapsed: 4269 ms, Size: 25.88 mb
Gzip,DeltaBinaryPacked,False Elapsed: 6365 ms, Size: 50.75 mb
Brotli,DeltaBinaryPacked,True Elapsed: 5112 ms, Size: 25.12 mb
Brotli,DeltaBinaryPacked,False Elapsed: 9378 ms, Size: 46.59 mb
Zstd,DeltaBinaryPacked,True Elapsed: 513 ms, Size: 31.45 mb
Zstd,DeltaBinaryPacked,False Elapsed: 771 ms, Size: 59.14 mb
Lz4,DeltaBinaryPacked,True Elapsed: 392 ms, Size: 47.44 mb
Lz4,DeltaBinaryPacked,False Elapsed: 506 ms, Size: 81.27 mb
Lz4Hadoop,DeltaBinaryPacked,True Elapsed: 390 ms, Size: 47.44 mb
Lz4Hadoop,DeltaBinaryPacked,False Elapsed: 506 ms, Size: 81.27 mb
Uncompressed,DeltaLengthByteArray,False Elapsed: 526 ms, Size: 106.63 mb
Snappy,DeltaLengthByteArray,False Elapsed: 579 ms, Size: 81.15 mb
Gzip,DeltaLengthByteArray,False Elapsed: 6366 ms, Size: 50.75 mb
Brotli,DeltaLengthByteArray,False Elapsed: 9347 ms, Size: 46.59 mb
Zstd,DeltaLengthByteArray,False Elapsed: 606 ms, Size: 59.14 mb
Lz4,DeltaLengthByteArray,False Elapsed: 647 ms, Size: 81.27 mb
Lz4Hadoop,DeltaLengthByteArray,False Elapsed: 654 ms, Size: 81.27 mb
Uncompressed,DeltaByteArray,False Elapsed: 402 ms, Size: 106.63 mb
Snappy,DeltaByteArray,False Elapsed: 577 ms, Size: 81.15 mb
Gzip,DeltaByteArray,False Elapsed: 6350 ms, Size: 50.75 mb
Brotli,DeltaByteArray,False Elapsed: 9168 ms, Size: 46.59 mb
Zstd,DeltaByteArray,False Elapsed: 728 ms, Size: 59.14 mb
Lz4,DeltaByteArray,False Elapsed: 501 ms, Size: 81.27 mb
Lz4Hadoop,DeltaByteArray,False Elapsed: 490 ms, Size: 81.27 mb
Uncompressed,ByteStreamSplit,False Elapsed: 381 ms, Size: 106.63 mb
Snappy,ByteStreamSplit,False Elapsed: 697 ms, Size: 81.15 mb
Gzip,ByteStreamSplit,False Elapsed: 6339 ms, Size: 50.75 mb
Brotli,ByteStreamSplit,False Elapsed: 9477 ms, Size: 46.59 mb
Zstd,ByteStreamSplit,False Elapsed: 591 ms, Size: 59.14 mb
Lz4,ByteStreamSplit,False Elapsed: 497 ms, Size: 81.27 mb
Lz4Hadoop,ByteStreamSplit,False Elapsed: 496 ms, Size: 81.27 mb
```

Some of the best results are:
* Snappy,DeltaBinaryPacked,True Elapsed: 597 ms, Size: 39.25 mb
* Zstd,DeltaBinaryPacked,True Elapsed: 513 ms, Size: 31.45 mb

This is of course much better result than what I had before, but I would really like to know where the difference comes from :)

After checking the meta again I can see that the differences are the following (left: `parquet.net`, right: `parquetSharp`) (both for `Snappy` compression)

![image](https://github.com/G-Research/ParquetSharp/assets/20644772/2bca884b-8531-4aaf-9816-d64ce97791b1)

![image](https://github.com/G-Research/ParquetSharp/assets/20644772/748dee43-78f3-4fb3-a85d-d6ca6813a8f4)

![image](https://github.com/G-Research/ParquetSharp/assets/20644772/a047e817-835a-4b4d-853d-afca53b05c52)

see that the encoding section on the left contains the values 3 4 and 5 while the right side contains only 5 and 3. Does that mean that `parquet.net` combines some additional encodings?

according to the [specification](https://parquet.apache.org/docs/file-format/data-pages/encodings/): 

* Run Length Encoding / Bit-Packing Hybrid (RLE = 3)
* Bit-packed (Deprecated) (BIT_PACKED = 4)
* Delta Encoding (DELTA_BINARY_PACKED = 5)

I am not sure if these are the same flags from the meta file that we see on the screenshots.. Could you please check the my response once more and see where the diff is coming from?

For anyone interested in code:

<details>
  <summary>lots of code</summary>

```csharp
        private async void button1_Click(object sender, EventArgs e)
        {
            var dataMP = GetData();
            await Do3(dataMP);
        }

        public async Task Do3(SingleCandlestickMP[] dataMP)
        {
            await Do2(dataMP, encoding: Encoding.Plain);
            await Do2(dataMP, encoding: Encoding.PlainDictionary);
            await Do2(dataMP, encoding: Encoding.Rle);
            await Do2(dataMP, encoding: Encoding.BitPacked);
            await Do2(dataMP, encoding: Encoding.DeltaBinaryPacked);
            await Do2(dataMP, encoding: Encoding.DeltaLengthByteArray);
            await Do2(dataMP, encoding: Encoding.DeltaByteArray);
            await Do2(dataMP, encoding: Encoding.RleDictionary);
            await Do2(dataMP, encoding: Encoding.ByteStreamSplit);
        }

        public async Task Do2(SingleCandlestickMP[] dataMP, Encoding encoding)
        {
            await Do1(dataMP, compression: Compression.Uncompressed, encoding: encoding);
            await Do1(dataMP, compression: Compression.Snappy, encoding: encoding);
            await Do1(dataMP, compression: Compression.Gzip, encoding: encoding);
            await Do1(dataMP, compression: Compression.Brotli, encoding: encoding);
            await Do1(dataMP, compression: Compression.Zstd, encoding: encoding);
            await Do1(dataMP, compression: Compression.Lz4, encoding: encoding);
            await Do1(dataMP, compression: Compression.Lz4Frame, encoding: encoding);
            await Do1(dataMP, compression: Compression.Lzo, encoding: encoding);
            await Do1(dataMP, compression: Compression.Bz2, encoding: encoding);
            await Do1(dataMP, compression: Compression.Lz4Hadoop, encoding: encoding);
        }

        public async Task Do1(
            SingleCandlestickMP[] dataMP,
            Compression compression,
            Encoding encoding)
        {
            await Do0(dataMP, compression: compression, encoding: encoding, isDisableDictionary: true);
            await Do0(dataMP, compression: compression, encoding: encoding, isDisableDictionary: false);
        }

        public async Task Do0(
            SingleCandlestickMP[] dataMP,
            Compression compression,
            Encoding encoding,
            bool isDisableDictionary)
        {
            var adder = $"{compression},{encoding},{isDisableDictionary}";

            var builder = new ParquetSharp.WriterPropertiesBuilder();

            try
            {
                if (isDisableDictionary)
                    builder.DisableDictionary();

                builder.Compression(compression);
                builder.Encoding(encoding);
            }
            catch (Exception ex) {
                Console.WriteLine($"{adder} Exception: {ex.Message}");
                return;
            }

            await Do(dataMP, builder, adder);
        }

        public async Task Do(SingleCandlestickMP[] data, WriterPropertiesBuilder builder, string adder)
        {
            var filePath = Path.Combine(outputDirectory, $"{fileName}_{adder}.parquet");

            Console.Write($"{adder}");

            try
            {
                var sw = Stopwatch.StartNew();
                await data.SaveToParquet(filePath, builder);
                sw.Stop();
                var elapsed = sw.ElapsedMilliseconds;

                Console.WriteLine($" Elapsed: {elapsed} ms, Size: {Math.Round(new FileInfo(filePath).Length.ConvertBytesToMegabytes(), 2)} mb");
            }
            catch ( Exception ex )
            {
                Console.WriteLine($" Exception: {ex.Message}");
            }

            if (File.Exists(filePath))
                File.Delete(filePath);
        }

// note: this method is in extensions class
        public static async Task SaveToParquet(
            this SingleCandlestickMP[] data, 
            string fileFullPath, 
            WriterPropertiesBuilder builder)
        {
            var columns = new Column[]
            {
                new Column<long>("Id"),
                new Column<long>("OpenTime"),
                new Column<long>("Open"),
                new Column<long>("High"),
                new Column<long>("Low"),
                new Column<long>("Close"),
                new Column<long>("Volume"),
                new Column<long>("CloseTime")
            };

            var properties = builder.Build();

            using var writer = ParquetFile.CreateRowWriter<SingleCandlestickMP>(fileFullPath, properties, columns);
            writer.WriteRows(data);
            writer.Close();
        }
```
</details>

**Reply by pavlexander** [reply]
out of curiosity I ran the parquet processing again, for both libraries, using the `Zstd` compression:
* Parquet.net: 1106 ms, File size: 27.30514907836914 mb
* ParquetSharp: 772 ms, File size: 31.449615478515625 mb

Again, it seems like `parquest.net` takes the lead in file size race. But the processing time is worse. You can ignore this test, I think.. We can just stick to `Snappy` compression issue for now.

**Reply by marcin-krystianc** [reply]
Hi @pavlexander, thanks for this exhaustive testing.
I have a hunch that one more reason for ParquetSharp producing larger file sizes is the number of row groups. With  Parquet.NET you probably have only a single row group, whereas with ParquetSharp you will get multiple row groups. You can tune ParquetSharp writer settings to bump the maximum number of rows in a row group (`MaxRowGroupLength`) so you will have a single row group as well.

**Reply by pavlexander** [reply]
@marcin-krystianc  no, still not it :)

`parquetSharp`

```csharp
            var builder = new ParquetSharp.WriterPropertiesBuilder();
            builder.DisableDictionary();
            builder.Compression(Compression.Snappy);
            //builder.Compression(Compression.Zstd);
            builder.Encoding(ParquetSharp.Encoding.DeltaBinaryPacked);
            builder.MaxRowGroupLength(1_000_000);
            builder.Version(ParquetVersion.PARQUET_1_0);
            builder.DisableStatistics();
```

`parquet.net`

```csharp
            new ParquetSerializerOptions()
            {
                CompressionMethod = Parquet.CompressionMethod.Snappy, // Zstd
                RowGroupSize = 1_000_000,
                ParquetOptions = new ParquetOptions()
                {
                    UseDictionaryEncoding = false,
                    UseDeltaBinaryPackedEncoding = true,
                }
            }
```

results:
* parquet.net: 853 ms, File size: 31.26325035095215 mb
* parquetSharp: 526 ms, File size: 39.24139404296875 mb

**Reply by adamreeve** [reply]
Hi @pavlexander, I'm a little late to reply but the other setting that might reduce file sizes is increasing the data page size in the writer properties. This defaults to 1MB in ParquetSharp, and increasing it can help compression perform better if column data in a row group is much larger than this.

I haven't looked into how Parquet.Net currently behaves but this old issue suggests they only write a single page per column in a row group, which might explain the lower file size: https://github.com/aloneguid/parquet-dotnet/issues/137

**Reply by adamreeve** [reply]
I spent a bit of time looking into this and trying to replicate your results as I'm also interested to know what the difference is here. I found that if I generate integers uniformly between 0 and long.MaxValue then the file sizes I get are nearly identical, but if I generate values between 0 and int.MaxValue then the files produced by ParquetSharp are much larger than those produced by Parquet.Net. Although if I change the data type to int32 then the file sizes are nearly identical again.

When I inspected the contents of the files in the case where ParquetSharp generated larger files, I could see that uncompressed page sizes were much larger for ParquetSharp, although the data type, encoding and number of values in the page was exactly the same. This indicates to me that the difference is likely due to differences in the implementations of the delta binary packed encoding. I haven't looked into how this encoding works in detail, but it has parameters that affect its behaviour, so possibly the Arrow implementation could be better tuned to encode this data more compactly. These parameters are currently hard coded though so it's not possible to change them with ParquetSharp.

There was a comment about making these parameters configurable at https://github.com/apache/arrow/pull/34632/files#r1141643429, but it doesn't look like that has happened yet.

I don't know the distribution of your test data, my uniformly distributed data is probably not very realistic and real-world data might be encoded more efficiently, but I suspect this is the real reason you're still seeing a difference in file sizes.

**Reply by adamreeve** [reply]
I also just came across this change that was merged in Arrow recently that is supposed to reduce the encoded size significantly for the delta binary packed encoding in some cases: https://github.com/apache/arrow/pull/37940

It looks like that should hopefully be released in version 14.0

**Reply by adamreeve** [reply]
I did some quick testing and changing the `kValuesPerBlock` parameter didn't have much affect on file size, but the fix that was merged in https://github.com/apache/arrow/pull/37940 significantly decreases the size of files in my uniformly distributed test case, and brings the size in line with Parquet.Net. So it looks like once 14.0.0 is released we'll have much improved file sizes when using the delta binary packed encoding.

**Reply by pavlexander** [reply]
@adamreeve Hey! Thank you for your time checking on this issue! I would like to replicate your testing with arrow v14 on my own dataset. Could you please help me?

**edit1:** so far I have managed to build the `ParquetSharp` solution successfully after installing `cmake` and fixing the path env variable for `msbuild`. Using Windows. 

**edit2**: the next thing I did is downloaded an arrow source tagged with v14, then did the 
`cmake .. --preset ninja-release-minimal` followed by `cmake --build .`
the output is `arrow.dll`, `arrow.exp`, `arrow.lib`. I then removed a nuget reference in ParquetSharp project and tried adding a dll assembly dependency, but got an error:

> the reference is invalid or unsupported..

![image](https://github.com/G-Research/ParquetSharp/assets/20644772/411e5d97-bad8-488e-97a1-23ce0ecf6e67)

I don't really have much experience with interop, so if you could point me on obvious mistake - I would much appreciate it :)

Thank you again for help, regardless

**Reply by adamreeve** [reply]
Hi @pavlexander, sure, we statically link the ParquetSharp C++ wrapper to the Arrow library so you won't be able to test this by just linking a different dll unfortunately. The way I tested this was copy the Arrow port from vcpkg (https://github.com/microsoft/vcpkg/tree/master/ports/arrow) and update it to point to a recent arrow commit, and use this as a vcpkg "overlay-port" in my local code. I've put my changes to test arrow 14 in a branch on this repository (https://github.com/G-Research/ParquetSharp/tree/arrow-14-test) so you could try rebuilding from that branch. But by pushing it to this repo our CI has built a NuGet package you can download and use for testing which is probably easiest, look for the "nuget-package" link under artifacts at: https://github.com/G-Research/ParquetSharp/actions/runs/6500781100

**Reply by pavlexander** [reply]
that's perfect! Used the nuget package you provided.

As a reminder - here are the old results:
* Snappy,DeltaBinaryPacked,True Elapsed: 597 ms, Size: 39.25 mb
* Zstd,DeltaBinaryPacked,True Elapsed: 513 ms, Size: 31.45 mb

new results (ParquetSharp 13.0 beta-1 is what I believe the Nuget package version was called)
* Snappy,DeltaBinaryPacked,True Elapsed: 416 ms, Size: 31.04 mb
* Zstd,DeltaBinaryPacked,True Elapsed: 301 ms, Size: 29.99 mb

here is the final comparison of previous and new results

### snappy
* (prev) parquet.net: 853 ms, File size: 31.26325035095215 mb
* (prev) parquetSharp: 526 ms, File size: 39.24139404296875 mb
* (new) ParquetSharp: Elapsed: 416 ms, Size: 31.04 mb

### zstd
* (prev) Parquet.net: 1106 ms, File size: 27.30514907836914 mb
* (prev) ParquetSharp: 772 ms, File size: 31.449615478515625 mb
* (new) ParquetSharp: Elapsed: 301 ms, Size: 29.99 mb

## in conclusion
Snappy compression with the new arrow version is going to be better, just on par with parquet.net. HOWEVER, it looks like the performance is twice as good. This is a huge improvement. Note that I am running apps in debug mode during testing, so the results might differ in release mode and with different datasets.

In regards to Zstd compression I see that parquet.net is slightly more effective, but is it worth 4x processing time? :)

@adamreeve,
I will keep this topic active for a bit and wait until the new version arrives, so I can re-test.. Unless I forget about it then feel free to close it as you wish :) Thank you very much for help, my curiosity has been fulfilled.

@marcin-krystianc much appreciate your help as well!
